[Bug] BLOB inline fields configured by blob-descriptor-field or blob-view-field require values to be a BlobDescriptor or BlobViewStruct.

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.


### Paimon version

1.5-SNAPSHOT

### Compute Engine

flink 1.20

### Minimal reproduce step

```
CREATE TABLE test_blob (
    id STRING,
    content BYTES
) WITH (
    'blob-field' = 'content',
    'blob-descriptor-field' = 'content'
);

INSERT INTO test_blob
/*+ OPTIONS('sink.parallelism' = '2') */
VALUES (
    'row1',
    sys.path_to_descriptor('file:///tmp/test.blob')
);
```


### What doesn't meet your expectations?

```
java.io.IOException: java.lang.IllegalArgumentException: BLOB inline fields configured by blob-descriptor-field or blob-view-field require values to be a BlobDescriptor or BlobViewStruct.
	at org.apache.paimon.flink.sink.RowDataStoreWriteOperator.write(RowDataStoreWriteOperator.java:63)
	at org.apache.paimon.flink.sink.RowDataStoreWriteOperator.processElement(RowDataStoreWriteOperator.java:55)
	at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask$StreamTaskNetworkOutput.emitRecord(OneInputStreamTask.java:246)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.processElement(AbstractStreamTaskNetworkInput.java:217)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.emitNext(AbstractStreamTaskNetworkInput.java:169)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:68)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:706)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:231)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:1202)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:1146)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:976)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:955)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:768)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:580)
	at java.base/java.lang.Thread.run(Thread.java:991)
Caused by: java.lang.IllegalArgumentException: BLOB inline fields configured by blob-descriptor-field or blob-view-field require values to be a BlobDescriptor or BlobViewStruct.
	at org.apache.paimon.format.parquet.writer.ParquetRowDataWriter$BlobDescriptorWriter.writeBlob(ParquetRowDataWriter.java:327)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataWriter$BlobDescriptorWriter.write(ParquetRowDataWriter.java:314)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataWriter$RowWriter.write(ParquetRowDataWriter.java:587)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataWriter.write(ParquetRowDataWriter.java:82)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataBuilder$ParquetWriteSupport.write(ParquetRowDataBuilder.java:86)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataBuilder$ParquetWriteSupport.write(ParquetRowDataBuilder.java:68)
	at org.apache.paimon.shade.org.apache.parquet.hadoop.ColumnCompressionRecordWriter.write(ColumnCompressionRecordWriter.java:157)
	at org.apache.paimon.shade.org.apache.parquet.hadoop.ParquetWriter.write(ParquetWriter.java:148)
	at org.apache.paimon.format.parquet.writer.ParquetBulkWriter.addElement(ParquetBulkWriter.java:53)
	at org.apache.paimon.io.SingleFileWriter.writeImpl(SingleFileWriter.java:146)
	at org.apache.paimon.io.StatsCollectingSingleFileWriter.write(StatsCollectingSingleFileWriter.java:61)
	at org.apache.paimon.io.RowDataFileWriter.write(RowDataFileWriter.java:140)
	at org.apache.paimon.io.RowDataFileWriter.write(RowDataFileWriter.java:50)
	at org.apache.paimon.io.RollingFileWriterImpl.write(RollingFileWriterImpl.java:77)
	at org.apache.paimon.utils.SinkWriter$DirectSinkWriter.write(SinkWriter.java:75)
	at org.apache.paimon.append.AppendOnlyWriter.write(AppendOnlyWriter.java:199)
	at org.apache.paimon.append.AppendOnlyWriter.write(AppendOnlyWriter.java:71)
	at org.apache.paimon.operation.AbstractFileStoreWrite.write(AbstractFileStoreWrite.java:177)
	at org.apache.paimon.table.sink.TableWriteImpl.writeAndReturn(TableWriteImpl.java:191)
	at org.apache.paimon.table.sink.TableWriteImpl.writeAndReturn(TableWriteImpl.java:179)
	at org.apache.paimon.flink.sink.StoreSinkWriteImpl.write(StoreSinkWriteImpl.java:108)
	at org.apache.paimon.flink.sink.RowDataStoreWriteOperator.write(RowDataStoreWriteOperator.java:61)
	... 14 more
Caused by: java.lang.RuntimeException: Blob data can not convert to descriptor.
	at org.apache.paimon.data.BlobData.toDescriptor(BlobData.java:53)
	at org.apache.paimon.data.Blob.serializeBlob(Blob.java:136)
	at org.apache.paimon.format.parquet.writer.ParquetRowDataWriter$BlobDescriptorWriter.writeBlob(ParquetRowDataWriter.java:325)
	... 35 more
```

### Anything else?

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/paimon-common/src/main/java/org/apache/paimon/data/AbstractBinaryWriter.java
+++ b/paimon-common/src/main/java/org/apache/paimon/data/AbstractBinaryWriter.java
@@ -203,7 +203,12 @@ abstract class AbstractBinaryWriter implements BinaryWriter {
 
     @Override
     public void writeBlob(int pos, Blob blob) {
-        byte[] bytes = blob.toData();
+        byte[] bytes;
+        if (blob instanceof BlobData) {
+            bytes = blob.toData();
+        } else {
+            bytes = Blob.serializeBlob(blob);
+        }
         writeBinary(pos, bytes, 0, bytes.length);
     }
 

--- a/paimon-common/src/main/java/org/apache/paimon/data/serializer/BlobSerializer.java
+++ b/paimon-common/src/main/java/org/apache/paimon/data/serializer/BlobSerializer.java
@@ -39,12 +39,18 @@ public class BlobSerializer extends SerializerSingleton<Blob> {
 
     @Override
     public void serialize(Blob blob, DataOutputView target) throws IOException {
-        BinarySerializer.INSTANCE.serialize(blob.toData(), target);
+        byte[] bytes;
+        if (blob instanceof BlobData) {
+            bytes = blob.toData();
+        } else {
+            bytes = Blob.serializeBlob(blob);
+        }
+        BinarySerializer.INSTANCE.serialize(bytes, target);
     }
 
     @Override
     public Blob deserialize(DataInputView source) throws IOException {
         byte[] bytes = BinarySerializer.INSTANCE.deserialize(source);
-        return new BlobData(bytes);
+        return Blob.fromBytes(bytes, null, null);
     }
 }

--- a/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/BlobTableITCase.java
+++ b/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/BlobTableITCase.java
@@ -724,6 +724,89 @@ public class BlobTableITCase extends CatalogITCaseBase {
         assertThat(AbstractFlinkTableFactory.schemaEquals(convertedRowType, flinkRowType)).isTrue();
     }
 
+    @Test
+    public void testBlobDescriptorFieldStreamingWriteWithShuffle() throws Exception {
+        // Verifies that blob-descriptor-field works when shuffle serialization occurs
+        // in streaming mode. When source and sink have different parallelism, Flink
+        // inserts a REBALANCE that serializes InternalRow to BinaryRow via
+        // InternalRowTypeSerializer. writeBlob() must preserve the BlobDescriptor so
+        // that BlobDescriptorWriter can serialize it at the sink.
+
+        byte[] blobData = "multimodal-content".getBytes();
+        FileIO fileIO = new LocalFileIO();
+        String uri = "file://" + warehouse + "/multimodal_blob_stream";
+        try (OutputStream outputStream =
+                fileIO.newOutputStream(new org.apache.paimon.fs.Path(uri), true)) {
+            outputStream.write(blobData);
+        }
+
+        sEnv.executeSql(
+                "CREATE TABLE multimodal_assets_metadata ("
+                        + "id STRING, "
+                        + "source_uri STRING, "
+                        + "media_content BYTES, "
+                        + "dt STRING"
+                        + ") PARTITIONED BY (dt) WITH ("
+                        + "'row-tracking.enabled'='true',"
+                        + "'data-evolution.enabled'='true',"
+                        + "'blob-field'='media_content',"
+                        + "'blob-descriptor-field'='media_content'"
+                        + ")");
+
+        // Streaming INSERT with sys.path_to_descriptor
+        // The /*+ OPTIONS('sink.parallelism' = '2') */ hint forces a shuffle
+        // from the source (parallelism 1 for VALUES) to the sink (parallelism 2).
+        // After fix, writeBlob() preserves BlobDescriptor through the shuffle.
+        sEnv.getConfig().set("table.dml-sync", "true");
+        sEnv.executeSql(
+                        "INSERT INTO multimodal_assets_metadata "
+                                + "/*+ OPTIONS('sink.parallelism' = '2') */ "
+                                + "VALUES ('test-id', 'test-uri', "
+                                + "sys.path_to_descriptor('"
+                                + uri
+                                + "'), '2025-07-07')")
+                .await();
+    }
+
+    @Test
+    public void testBlobDescriptorFieldBatchWriteWithShuffle() throws Exception {
+        // Same as streaming test but in batch mode.
+        // Verifies that blob-descriptor-field works when shuffle serialization
+        // occurs in batch mode.
+
+        byte[] blobData = "batch-multimodal-content".getBytes();
+        FileIO fileIO = new LocalFileIO();
+        String uri = "file://" + warehouse + "/multimodal_blob_batch";
+        try (OutputStream outputStream =
+                fileIO.newOutputStream(new org.apache.paimon.fs.Path(uri), true)) {
+            outputStream.write(blobData);
+        }
+
+        tEnv.executeSql(
+                "CREATE TABLE batch_multimodal_metadata ("
+                        + "id STRING, "
+                        + "source_uri STRING, "
+                        + "media_content BYTES, "
+                        + "dt STRING"
+                        + ") PARTITIONED BY (dt) WITH ("
+                        + "'row-tracking.enabled'='true',"
+                        + "'data-evolution.enabled'='true',"
+                        + "'blob-field'='media_content',"
+                        + "'blob-descriptor-field'='media_content'"
+                        + ")");
+
+        // Batch INSERT with sys.path_to_descriptor
+        // The /*+ OPTIONS('sink.parallelism' = '2') */ hint forces a shuffle,
+        // same as streaming mode.
+        batchSql(
+                "INSERT INTO batch_multimodal_metadata "
+                        + "/*+ OPTIONS('sink.parallelism' = '2') */ "
+                        + "VALUES ('test-id', 'test-uri', "
+                        + "sys.path_to_descriptor('"
+                        + uri
+                        + "'), '2025-07-07')");
+    }
+
     private static final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();
 
     public static String bytesToHex(byte[] bytes) {

